test/Test/BaseDAO.hs

```diff
@@ -0,0 +1,271 @@
+          , decimals = 8
+          , extras = Map.empty
+          }
+      , FA2.TokenMetadata
+          { token_id = 1
+          , symbol = [mt|frozen_token|]
+          , name = [mt|Frozen Token|]
+          , decimals = 8
+          , extras = Map.empty
+          }
+      ]
+
+-- TODO: Replace SampleFA2 with BaseDAO
+baseDAOContract :: FA2.PermissionsDescriptor -> Contract FA2.Parameter SampleFA2.Storage
+baseDAOContract p =
+  SampleFA2.fa2Contract p
+
+originateBaseDAO
+  :: MonadNettest caps base m => FA2.PermissionsDescriptor
+  -> m ((Address, Address), (Address, Address), TAddress FA2.Parameter, Address)
+originateBaseDAO p = do
+  owner1 :: Address <- newAddress "owner1"
+  operator1 :: Address <- newAddress "operator1"
+  owner2 :: Address <- newAddress "owner2"
+  operator2 :: Address <- newAddress "operator2"
+
+  admin :: Address <- newAddress "admin"
+
+  let bal = Map.fromList
+        [ ((owner1, 0), 100)
+        , ((owner2, 0), 100)
+        ]
+  let (operators :: SampleFA2.Operators) =
+        BigMap $ Map.fromList
+          [ ((owner1, operator1), ())
+          , ((owner2, operator2), ())
+          ]
+  let
+    originateData = OriginateData
+      { odFrom = AddressResolved genesisAddress
+      , odName = "BaseDAO"
+      , odBalance = toMutez 0
+      , odStorage = baseDAOStorage admin bal operators
+      , odContract = baseDAOContract p
+      }
+  dao <- originate originateData
+  pure ((owner1, operator1), (owner2, operator2), dao, admin)
+
+test_BaseDAO :: TestTree
+test_BaseDAO = testGroup "tests to check BaseDAO contract functionality"
+  [ testGroup "Operator:"
+      [ testCase "allows zero transfer from non-existent operator" $
+          nettestTestExpectation zeroTransferScenario
+      , testCase "aborts if there is a failure (due to low balance)" $
+          nettestTestExpectation lowBalanceScenario
+      , testCase "aborts if there is a failure (due to non existent source account)" $
+          nettestTestExpectation noSourceAccountScenario
+      , testCase "aborts if there is a failure (due to bad operator)" $
+          nettestTestExpectation badOperatorScenario
+      , testCase "accepts an empty list of transfers" $
+          nettestTestExpectation emptyTransferListScenario
+      , testCase "validates token id" $
+          nettestTestExpectation validateTokenScenario
+      , testCase "cannot transfer foreign money" $
+          nettestTestExpectation noForeignMoneyScenario
+      ]
+  , testGroup "Owner:"
+      [ testCase "aborts if there is a failure (due to low balance)" $
+          nettestTestExpectation lowBalanceOwnerScenario
+      , testCase "validates token id" $
+          nettestTestExpectation validateTokenOwnerScenario
+      , testCase "cannot transfer foreign money" $
+          nettestTestExpectation noForeignMoneyOwnerScenario
+      , testCase "balanceOf request validates token id" $
+          nettestTestExpectation balanceOfOwnerScenario
+      ]
```

Comment:
It looks like the only *positive* test is zero transfer, all other tests check incorrect scenarios when something is expected to fail. Can you add tests for most common cases: when there is valid transfer and we expect it to update balances properly.

Reply:
I've added the `validTransfer` cases, and `updateOperator` cases. Let me know if there are other cases that I should add.